### PR TITLE
Clean future goals

### LINKED ISSUE #10939: Anomaly: Evars generated outside of proof engine...
#### Description of the problem

~~~coq
Goal False.
  epose (ltac:(
    let x := fresh in 
    intros x;
    induction x; admit) : forall pf : 7=7, True).
(*
Anomaly
"Evars generated outside of proof engine (e.g. V82, clear, ...) are not supposed to be explicitly given up."
Please report at http://coq.inria.fr/bugs/.
*)
~~~

#### Coq Version

8.9.1

### PR DESCRIPTION
**Pull request by maximedenes**
We make the `future_goals` stack explicit in the evarmap. This is a preliminary change before moving the shelf to the evarmap, which is necessary for #7825 and other changes where shelving info is needed by components which are not in the tactic monad.

We also move given up goals to the evarmap, which makes it possible to fix a long standing issue triggering an anomaly.

Fixes https://github.com/coq/coq/issues/10939

### CODE REVIEW

**Comment by maximedenes** [comment]
@ppedrot would you mind being the assignee? We already discussed the context of this PR some time ago.

File: proofs/proof.ml
Comment (maximedenes) [reply proofs/proof.ml]: You're right, I failed to do it initially because I hadn't yet understood #12878. I now rebased on top and put push/pop back in order.

File: engine/evd.mli
Comment (maximedenes) [reply engine/evd.mli]: Fixed, thanks.

**Comment by maximedenes** [comment]
@coq/ci-maintainers any idea where the windows64 timeout could from? It's the first time I see it, and the build took only 12 minutes...

**Comment by SkySkimmer** [comment]
No idea

**Comment by maximedenes** [comment]
@ppedrot All green, this should be ready for merge. Next in line is #12872 :)

File: engine/evd.ml
Comment (ppedrot) [inline engine/evd.ml:467]: Is it possible to make this type private? It doesn't seem to be created outside of this API.

File: engine/evd.mli
Comment (ejgallego) [inline engine/evd.mli:367]: Maybe it'd be worth making `future_goals` a module?

**Comment by maximedenes** [comment]
@ppedrot does it look good apart from that nit? I'd rather address all concerns before relaunching a pipeline.

**Comment by ppedrot** [comment]
At first sight everything looks fine, but it's the kind of super-duper tricky code that has far-reaching effects. The CI seems fine, so that's a good enough empirical observation for me.

File: engine/evd.ml
Comment (mattam82) [inline engine/evd.ml:580]: I guess the comment is not relevant anymore.

File: proofs/proof.ml
Comment (mattam82) [inline proofs/proof.ml:373]: Not different from before but, why is this simply "undefined" and not up-to advance? 
If a future goal is instantiated by another new evar, I guess the other evar should be a future goal too. It could also be an older goal, and we don't want to consider this possibility here. Is that reasoning correct?

**Review by mattam82 — APPROVED** [review]
Looks good on my side, I much prefer the push/pop interface!

File: proofs/proof.ml
Comment (maximedenes) [reply proofs/proof.ml:373]: Aren't all new evars (until they are instantiated) declared as future goals?

**Comment by maximedenes** [comment]
@ppedrot Comments addressed, all green!

**Review by ppedrot — APPROVED** [review]
Good enough for me.

**Comment by ppedrot** [comment]
Final round of comments before I merge.

**Comment by maximedenes** [comment]
@ppedrot Thanks!

**Pull request merged**
